File: DotNet/LotsOfSatellites/Satellites.cs
using System.Collections.Generic;
using AGI.Foundation;
using AGI.Foundation.Coordinates;
using AGI.Foundation.Graphics;
using AGI.Foundation.Time;

namespace AGI.Examples.LotsOfSatellites
{
    public class Satellites
    {
        public Satellites()
        {
            m_satellites = new List<MotionEvaluator<Cartesian>>();
            m_accessPositions = new List<Cartesian>();
            m_noAccessPositions = new List<Cartesian>();

            m_accessBatch = new MarkerBatchPrimitive
            {
                Texture = SceneManager.Textures.FromUri(LotsOfSatellites.GetDataFilePath("Markers/Satellite-Red.png"))
            };

            m_noAccessBatch = new MarkerBatchPrimitive
            {
                Texture = SceneManager.Textures.FromUri(LotsOfSatellites.GetDataFilePath("Markers/Satellite.png"))
            };

            SceneManager.Primitives.Add(m_accessBatch);
            SceneManager.Primitives.Add(m_noAccessBatch);
        }

        public void Add(MotionEvaluator<Cartesian> evaluator, JulianDate epoch)
        {
            Cartesian position = evaluator.Evaluate(epoch);
            m_noAccessPositions.Add(position);

            m_satellites.Add(evaluator);
        }

        public void AppendPosition(Cartesian position, bool showAccess)
        {
            if (showAccess)
            {
                m_accessPositions.Add(position);
            }
            else
            {
                m_noAccessPositions.Add(position);
            }
        }

        public void SetMarkerBatches()
        {
            if (m_accessPositions.Count == 0)
            {
                m_accessPositions.Add(Cartesian.Zero);
            }

            if (m_noAccessPositions.Count == 0)
            {
                m_noAccessPositions.Add(Cartesian.Zero);
            }

            m_accessBatch.Set(m_accessPositions);
            m_noAccessBatch.Set(m_noAccessPositions);
        }

        public void ClearPositions()
        {
            m_accessPositions.Clear();
            m_noAccessPositions.Clear();
        }

        public MotionEvaluator<Cartesian> GetSatellite(int index)
        {
            return m_satellites[index];
        }

        public int Count
        {
            get { return m_satellites.Count; }
        }

        public void RemoveUsingIndexList(List<int> satellitesToRemove)
        {
            for (int i = satellitesToRemove.Count - 1; i >= 0; --i)
            {
                m_satellites.RemoveAt(satellitesToRemove[i]);
            }
        }

        public void Clear()
        {
            m_satellites.Clear();
            m_accessPositions.Clear();
            m_noAccessPositions.Clear();
        }

        public void ClearAccesses()
        {
            foreach (Cartesian position in m_accessPositions)
            {
                m_noAccessPositions.Add(position);
            }

            m_accessPositions.Clear();
        }

        private readonly List<MotionEvaluator<Cartesian>> m_satellites;
        private readonly List<Cartesian> m_accessPositions;
        private readonly List<Cartesian> m_noAccessPositions;
        private readonly MarkerBatchPrimitive m_accessBatch;
        private readonly MarkerBatchPrimitive m_noAccessBatch;
    }
}

File: Java/LotsOfSatellites/src/agi/examples/lotsofsatellites/Satellites.java
package agi.examples.lotsofsatellites;

import java.util.ArrayList;

import agi.foundation.MotionEvaluator1;
import agi.foundation.coordinates.Cartesian;
import agi.foundation.graphics.MarkerBatchPrimitive;
import agi.foundation.graphics.SceneManager;
import agi.foundation.time.JulianDate;

public class Satellites {
    public Satellites() {
        m_satellites = new ArrayList<>();
        m_accessPositions = new ArrayList<>();
        m_noAccessPositions = new ArrayList<>();

        m_accessBatch = new MarkerBatchPrimitive();
        m_accessBatch.setTexture(SceneManager.getTextures().fromUri(LotsOfSatellites.getDataFilePath("Markers/Satellite-Red.png")));

        m_noAccessBatch = new MarkerBatchPrimitive();
        m_noAccessBatch.setTexture(SceneManager.getTextures().fromUri(LotsOfSatellites.getDataFilePath("Markers/Satellite.png")));

        SceneManager.getPrimitives().add(m_accessBatch);
        SceneManager.getPrimitives().add(m_noAccessBatch);
    }

    public void add(MotionEvaluator1<Cartesian> evaluator, JulianDate epoch) {
        Cartesian position = evaluator.evaluate(epoch);
        m_noAccessPositions.add(position);
        m_satellites.add(evaluator);
    }

    public void appendPosition(Cartesian position, boolean showAccess) {
        if (showAccess) {
            m_accessPositions.add(position);
        } else {
            m_noAccessPositions.add(position);
        }
    }

    public void setMarkerBatches() {
        if (m_accessPositions.size() == 0) {
            m_accessPositions.add(Cartesian.getZero());
        }

        if (m_noAccessPositions.size() == 0) {
            m_noAccessPositions.add(Cartesian.getZero());
        }

        m_accessBatch.set(m_accessPositions);
        m_noAccessBatch.set(m_noAccessPositions);
    }

    public final void clearPositions() {
        m_accessPositions.clear();
        m_noAccessPositions.clear();
    }

    public final MotionEvaluator1<Cartesian> getSatellite(int index) {
        return m_satellites.get(index);
    }

    public final int getCount() {
        return m_satellites.size();
    }

    public final void removeUsingIndexList(ArrayList<Integer> satellitesToRemove) {
        for (int i = satellitesToRemove.size() - 1; i >= 0; --i) {
            m_satellites.remove((int)satellitesToRemove.get(i));
        }
    }

    public final void clear() {
        m_satellites.clear();
        m_accessPositions.clear();
        m_noAccessPositions.clear();
    }

    public final void clearAccesses() {
        for (Cartesian position : m_accessPositions) {
            m_noAccessPositions.add(position);
        }

        m_accessPositions.clear();
    }

    private final ArrayList<MotionEvaluator1<Cartesian>> m_satellites;
    private final ArrayList<Cartesian> m_accessPositions;
    private final ArrayList<Cartesian> m_noAccessPositions;
    private final MarkerBatchPrimitive m_accessBatch;
    private final MarkerBatchPrimitive m_noAccessBatch;
}